plugins/source/hackernews/resources/services/items/items_fetch.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/hermanschaaf/hackernews"
+	"golang.org/x/sync/errgroup"
+)
+
+// fetchItems gets called by the CloudQuery SDK to fetch the data for the `hackernews_items` table. It is passed a context and a
+// schema.ClientMeta interface which is used to access the plugin configuration and the state backend. The
+// resource argument is not used here, as it refers to the parent table, which is not applicable in this case.
+// The res channel is used to send the fetched data back to the CloudQuery SDK.
+//
+// Because this is an incremental table, we must first load the cursor from the state backend, and then update
+// it whenever we've successfully fetched a new item. The cursor is a string, and in this case it is the ID of
+// the last item we've fetched.
+//
+// Incremental tables should guarantee at-least-once delivery. The Hacker News API gives us the current maximum item ID,
+// and we need to fetch all items up to that ID. Once we've fetched all items up to a certain ID, we can safely
+// update the cursor. If we crash before updating the cursor, we will fetch the same items again on the next run.
+//
+// This resolver is different from most in that it needs to concurrently fetch items using its own goroutines. Usually,
+// a resolver with pagination will not have this ability and will simply iterate over all pages.
+func fetchItems(ctx context.Context, meta schema.ClientMeta, _ *schema.Resource, res chan<- any) error {
+	c := meta.(*client.Client)
+	tableName := Items().Name
+	value, err := c.Backend.Get(ctx, tableName, "id")
+	if err != nil {
+		return fmt.Errorf("failed to retrieve state from backend: %w", err)
+	}
+
+	// read the cursor from the state, or default to 0 if it's not set
+	cursor := 0
+	if value != "" {
+		cursor, err = strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf("failed to convert cursor to int: %w", err)
+		}
+	}
+	c.Logger().Info().Msg("Found previous cursor, starting from " + strconv.Itoa(cursor))
+
+	// find the max item ID from the Hacker News API
+	work := make(chan int)
+	maxID, err := c.HackerNews.MaxItemID(ctx)
+	if err != nil {
+		return err
+	}
+	c.Logger().Info().Msg("Found max ID, reading up to " + strconv.Itoa(maxID))
+
+	// spin up worker goroutines that will fetch items concurrently by reading ids from
+	// the work channel
+	workers := c.Spec.ItemConcurrency
+	success := make(chan int)
+	g, gctx := errgroup.WithContext(ctx)
+	for i := 0; i < workers; i++ {
+		g.Go(func() error {
+			for {
+				select {
+				case <-gctx.Done():
+					return gctx.Err()
+				case itemID, ok := <-work:
+					if !ok {
+						return nil
+					}
+					fetchErr := c.RetryOnError(ctx, tableName, func() error {
+						return fetchItem(gctx, c, itemID, res)
+					})
+					if fetchErr != nil {
+						return fetchErr
+					}
+					success <- itemID
+				}
+			}
+		})
+	}
+
+	// We use a min heap to keep track of the items that have finished fetching,
```

Comment:
You could try to put this in another file, grouped under worker/manager 'structs' and divided into multiple methods.

Reply:
I'm saving this response to use in future code reviews on my own PRs.

Reply:
Haha 😂  The thing is, there's a lot going on here. Breaking it up into smaller functions may make each function simpler, but the overall system harder to understand. I'll see what I can do though

Reply:
I did a refactor now but I'm not sure it's better, and not sure it's still correct either. Haven't pushed it up yet, I'll need to verify it with fresh eyes tomorrow